### DISCUSSION TITLE
Right way to configure LSP?

### DISCUSSION THREAD

**Discussion by alewis001** [LazyVim/LazyVim#2407] [Q&A] [answered]
In the comment within the `setup` section of https://www.lazyvim.org/plugins/lsp#nvim-lspconfig it says:

```
 setup = {
    -- example to setup with typescript.nvim
    -- tsserver = function(_, opts)
    --   require("typescript").setup({ server = opts })
    --   return true
    -- end,
    -- Specify * to use this function as a fallback for any server
    -- ["*"] = function(server, opts) end,
  },
```

I tried this with Volar such that I had...

```
  {
    "nvim-lspconfig",
    opts = {
      setup = {
        volar = function(_, opts)
          require("volar").setup({
            on_attach = function(client)
              client.server_capabilities.documentFormattingProvider = false
            end,
          })
        end,
      },
    },
  },

```
However, on nvim startup I get a module not found for the `require("volar")` line. I also get a warning about the unused `opts` which worries me that even if I got the missing module sorted, I could still be missing something important with the `opts` param. 

Instead of the documented approach the following does appear to work:

```
  {
    "nvim-lspconfig",
    opts = {
      setup = {
        volar = function()
          require("lazyvim.util").lsp.on_attach(function(client, _)
            if client.name == "volar" then
              client.server_capabilities.documentFormattingProvider = false
            end
          end)
        end,
      },
    },
  }
```
Given the `if` in that solution it would suggest that the function is being called for every LSP that is setup, which I assume would be less desirable than directly configuring Volar directly which is my interpretation of what the code in the docs would do. The `volar = function()` would also suggest it would only be called for Volar so 🤷 

If it's not already abundantly clear, I'm a neovim and lazyvim noob so I appreciate your patience :)

Cheers

**Comment by dpetka2001** [comment, marked as answer]
That example you linked to, uses `typescript.nvim` that used to be part of LazyVim to setup the `tsserver` before it got archived. So, the `require("typescript")` that is used in the example is actually referring to the module of `typescript.nvim` that used to setup the `tsserver`. The correct way to do it in your case is the second example you mentioned.

**Reply by alewis001** [reply]
That's great. Thank you for the confirmation, and for your help.
